### ISSUE TITLE
Sounddevice doesn't play if a program that uses sound is opened.

### ISSUE THREAD

**Issue by OmarSquircleArt** [spatialaudio/python-sounddevice#81]
Hello

I am on Arch Linux and I started using SD today. My initial tests worked just fine. But when I opened another application like Spotify that play music (It is just opened but it doesn't play anything) and then tried to play a sound, it doesn't work anymore. Even if I closed Spotify it won't work till I restart my PC.

Devices are listed normally and it doesn't return any errors. Is this a normal behaviour or is it a bug?

And just a small question, Can I play during the playing of another sound? For instance, if a sound's duration is 1s, can I play the sound then after 0.5s play it again without the second sound killing the first sound (I want them to play at the same time).

Thanks.

**Comment by mgeier** [comment]
What about other applications that play sound? Do they work when Spotify is open?

You should check with some other application that uses PortAudio, e.g. Audacity.

**Comment by OmarSquircleArt** [comment]
I just tried Audacity and it seems to have similar problems. Either Spotify or Audacity will work.
And Audacity return this error when it doesn't work, SoundDevice returnes similar error:
`Error opening sound device. Try changing the audio host, playback device and the project sample rate.`

Other applications like Chromium will work normally if Spotify is opened.

Is this a bug in Port Audio?

**Comment by OmarSquircleArt** [comment]
I did some researching and it turned out to be a configuration issue. I configured ALSA and Dmix and all the applications worked fine.

~~~

   0 HDA Intel PCH: ALC887-VD Analog (hw:0,0), ALSA (2 in, 2 out)
   1 HDA Intel PCH: ALC887-VD Digital (hw:0,1), ALSA (0 in, 2 out)
   2 HDA Intel PCH: ALC887-VD Alt Analog (hw:0,2), ALSA (2 in, 0 out)
   3 HDA ATI HDMI: 0 (hw:1,3), ALSA (0 in, 8 out)
   4 sysdefault, ALSA (128 in, 128 out)
   5 front, ALSA (0 in, 2 out)
   6 surround40, ALSA (0 in, 2 out)
   7 surround51, ALSA (0 in, 2 out)
   8 surround71, ALSA (0 in, 2 out)
   9 iec958, ALSA (0 in, 2 out)
  10 spdif, ALSA (0 in, 2 out)
  11 pulse, ALSA (32 in, 32 out)
* 12 default, ALSA (128 in, 128 out)
  13 dmix, ALSA (0 in, 2 out)

~~~

SD automatically set ALSA as default I believe. However, sometimes ALSA and Dmix don't even show up in the list, not sure what the issue is here.

Thanks ! btw, is it possible to play 2 sound at the same time inside SD?

**Comment by tgarc** [comment]
@OmarSquircleArt If you're asking whether you can play multiple audio sources out of the same device simultaneously then what your describing is 'mixing'. This is not something that's built into portaudio or sounddevice. If you're asking whether it's possible to play out to multiple audio devices simultaneously I think this is possible.

**Comment by OmarSquircleArt** [comment]
@tgarc I am aware that it is mixing. I could mix my audio before playing it but this won't make it realtime. So is there is anyway mixing can be handled automatically in some way when 2 sounds are played in the same interval?

**Comment by tgarc** [comment]
automagically? No. You would have to mix it yourself using your own audio [callback](http://python-sounddevice.readthedocs.io/en/0.3.7/#callback-streams). I'd thought to add an example of this to the project but I haven't had/don't have the time right now. Depending on how extensive your mixing needs are, you may want to look at a library that is written for this. (I know pygame has bindings for SDLmixer for example). But if your mixing goals are modest it should be easy enough to achieve with sounddevice alone. Some assembly required :D

**Comment by OmarSquircleArt** [comment]
My mixing goals are modest yah. I just need the ability to play multiple samples together (A simple sum).
But assembly? No thank you :smile: I don't have much codding experience.
I will wait for your example, take your time.

**Comment by tgarc** [comment]
LoL you know I didn't mean assembly code right?

**Comment by OmarSquircleArt** [comment]
I suspected :laughing: , I really though you meant assembly code at first (because we are kinda going low level), but nah, nobody use assembly now.

**Comment by tgarc** [comment]
@OmarSquircleArt 
Your question has been in the back of my mind for a while and I was thinking, the simplest solution to your problem would probably to be rely on the mixing capabilities of the audio backend itself.
For example, here's something I whipped in a hurry that uses the default ALSA device provided in linux mint:

```python
from __future__ import print_function
import sounddevice as sd
from numpy import pi, sin, arange
import threading
import time

def streamer(stream, data):
    stream.start()
    stream.write(data)
    stream.close()

for i in range(1,7):
    x = 0.1*sin(2*pi*i*440*arange(96e3)/48e3, dtype='float32')
    stream = sd.OutputStream(device='default', samplerate=48000, channels=1, dtype='float32')
    thread = threading.Thread(target=streamer, args=(stream, x))
    thread.start()
    print('Sin %dHz' % (i*440))
    time.sleep(0.5)
time.sleep(5)
```

The tricky part is you cannot use the simple play/record functions; you have to write your own callbacks or use the blocking read/write functions with threads. Also, this only works if your backend device is configured for mixing (the 'default' alsa device in my example uses the [dmix](http://alsa.opensrc.org/Dmix) plugin).

NB my example is sloppy as heck, don't expect any robustness.

**Comment by OmarSquircleArt** [comment]
Thanks! Will test that soon and so some more studying of the subject.

**Comment by mgeier** [comment]
@OmarSquircleArt Any news?

**Comment by OmarSquircleArt** [comment]
@mgeier Sorry I didn't get back to you on time.
I did try this and it worked just fine for my tests. Thanks !

I used SD a while ago to make a completely procedural synthesizer, perhaps you would like to see it in action, here are some test I did:

<blockquote class="twitter-video" data-lang="en"><p lang="en" dir="ltr">All sounds you hear are synthesized procedurally using Animation Nodes in realtime. And played also in AN. <a href="https://t.co/NfzFuDkQJC">pic.twitter.com/NfzFuDkQJC</a></p>&mdash; Omar Ahmad (@omarsquircleart) <a href="https://twitter.com/omarsquircleart/status/849336903044390913?ref_src=twsrc%5Etfw">April 4, 2017</a></blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

**Comment by mgeier** [comment]
Thanks for the update, looks interesting! Is the source code available?

**Comment by OmarSquircleArt** [comment]
@mgeier It is not really "coded". This [system](https://github.com/JacquesLucke/animation_nodes) is a node based scripting system. That might sound weird, so here is an example to compute the sine wave samples you computed above:

![Example 1](https://user-images.githubusercontent.com/25300994/31903311-0f05816e-b828-11e7-8350-f336444e218c.png)

As you may see each function is implemented as a "node", what I did was make a node that takes those computed samples as input and play them using SoundDevice.

Loops and function can also be used, here is an example to make the same loop as above:

![Example 2](https://user-images.githubusercontent.com/25300994/31903830-b66e22ac-b829-11e7-992d-809bea47318c.png)

In conclusion, this system allows me to make cool stuff without having to worry about coding, and this is usually what students want.

**Comment by mgeier** [comment]
Cool, thanks for the info!

**Comment by tgarc** [comment]
@OmarSquircleArt  ohh this is really cool..thanks for sharing!

**Comment by SumZbrod** [comment]
How to solve a problem " No output device matching 'default'"

**Comment by HaHeho** [comment]
> How to solve a problem " No output device matching 'default'"

Check the available audio devices with `sd.query_devices()` and choose a valid name. In your case there seems to be no "default" available, as this may be system dependent (e.g. I have no "default" shown on Windows).
